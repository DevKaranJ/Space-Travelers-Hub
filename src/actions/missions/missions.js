import axios from 'axios';

export const fetchMissions = () => async (dispatch) => {
  try {
    const response = await axios.get('https://api.spacexdata.com/v3/missions');
    dispatch({ type: 'FETCH_MISSIONS_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({ type: 'FETCH_MISSIONS_FAILURE', payload: error.message });
  }
};

export const joinMission = (missionId) => ({ type: 'JOIN_MISSION', payload: missionId });

export const leaveMission = (missionId) => ({ type: 'LEAVE_MISSION', payload: missionId });
