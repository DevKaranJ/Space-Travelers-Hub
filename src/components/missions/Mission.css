.mission-table {
  width: 100%;
  border-collapse: collapse;
  margin-top: 20px;
}

.mission-table th, .mission-table td {
  border: 1px solid #ccc;
  padding: 8px;
  text-align: left;
}

.mission-table th {
  background-color: #f2f2f2;
}

.mission-table tr:hover {
  background-color: #f5f5f5;
}

.mission-table button {
  padding: 5px 10px;
  background-color: #007BFF;
  color: #fff;
  border: none;
  cursor: pointer;
}

.mission-table button:hover {
  background-color: #0056b3;
}
