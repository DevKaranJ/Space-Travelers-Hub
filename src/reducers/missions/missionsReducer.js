const initialState = {
  missions: [],
  error: null,
};

const missionsReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_MISSIONS_SUCCESS':
      return {
        ...state,
        missions: action.payload,
        error: null,
      };
    case 'FETCH_MISSIONS_FAILURE':
      return {
        ...state,
        missions: [],
        error: action.payload,
      };
    case 'JOIN_MISSION':
      return {
        ...state,
        missions: state.missions.map((mission) => {
          if (mission.mission_id === action.payload) {
            return { ...mission, reserved: true };
          }
          return mission;
        }),
      };
    case 'LEAVE_MISSION':
      return {
        ...state,
        missions: state.missions.map((mission) => {
          if (mission.mission_id === action.payload) {
            return { ...mission, reserved: false };
          }
          return mission;
        }),
      };
    default:
      return state;
  }
};

export default missionsReducer;
