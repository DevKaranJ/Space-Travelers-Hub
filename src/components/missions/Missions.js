import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions, joinMission, leaveMission } from '../../actions/missions/missions';

function Missions() {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);
  const error = useSelector((state) => state.missions.error);

  const handleJoinMission = (missionId) => {
    dispatch(joinMission(missionId));
  };

  const handleLeaveMission = (missionId) => {
    dispatch(leaveMission(missionId));
  };

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  if (error) {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <div>
      <h2>Missions</h2>
      <ul>
        {missions.map((mission) => (
          <li key={mission.mission_id}>
            <h3>{mission.mission_name}</h3>
            <p>{mission.description}</p>
            {mission.reserved ? (
              <div>
                <span>Active Member</span>
                <button type="button" onClick={() => handleLeaveMission(mission.mission_id)}>Leave Mission</button>
              </div>
            ) : (
              <div>
                <span>NOT A MEMBER</span>
                <button type="button" onClick={() => handleJoinMission(mission.mission_id)}>Join Mission</button>
              </div>
            )}
          </li>
        ))}
      </ul>
    </div>
  );
}

export default Missions;
