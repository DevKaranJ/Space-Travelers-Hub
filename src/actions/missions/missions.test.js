import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';
import { fetchMissions, joinMission, leaveMission } from './missions'; // Replace with the actual path to your actions

jest.mock('axios'); // Mock the axios module

const mockStore = configureMockStore([thunk]);

describe('Redux Actions', () => {
  let store;

  beforeEach(() => {
    store = mockStore({});
  });

  it('should create a snapshot for fetchMissions action', async () => {
    axios.get.mockResolvedValue({ data: [{ mission: 'Mission 1' }, { mission: 'Mission 2' }] });

    await store.dispatch(fetchMissions());

    expect(store.getActions()).toMatchSnapshot();
  });

  it('should create a snapshot for joinMission action', () => {
    const missionId = 1;
    store.dispatch(joinMission(missionId));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should create a snapshot for leaveMission action', () => {
    const missionId = 1;
    store.dispatch(leaveMission(missionId));
    expect(store.getActions()).toMatchSnapshot();
  });
});
