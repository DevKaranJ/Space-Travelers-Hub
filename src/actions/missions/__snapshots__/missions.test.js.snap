// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Actions should create a snapshot for fetchMissions action 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "mission": "Mission 1",
      },
      Object {
        "mission": "Mission 2",
      },
    ],
    "type": "FETCH_MISSIONS_SUCCESS",
  },
]
`;

exports[`Redux Actions should create a snapshot for joinMission action 1`] = `
Array [
  Object {
    "payload": 1,
    "type": "JOIN_MISSION",
  },
]
`;

exports[`Redux Actions should create a snapshot for leaveMission action 1`] = `
Array [
  Object {
    "payload": 1,
    "type": "LEAVE_MISSION",
  },
]
`;
